File: frontend/src/types/status.ts
import type { EntityConfig } from './domain';

export type GateState = 'open' | 'closed' | 'moving' | 'locked';
export const ALL_GATE_STATE: readonly GateState[] = ['open', 'closed', 'moving', 'locked'];
export type DirectiveState = 'draft' | 'pending' | 'approved' | 'executing' | 'completed' | 'aborted';
export const ALL_DIRECTIVE_STATE: readonly DirectiveState[] = ['draft', 'pending', 'approved', 'executing', 'completed', 'aborted'];

export const ENTITY_TRANSITIONS: Readonly<Record<string, Readonly<Record<string, readonly string[]>>>> = {
	reservoir: { normal: ['warning', 'critical'], warning: ['critical', 'restricted', 'normal'], critical: ['restricted', 'warning'], restricted: ['critical'] },
	gateUnit: { open: ['moving', 'locked'], closed: ['moving', 'locked'], moving: ['open', 'closed', 'locked'], locked: ['closed'] },
	operationDirective: { draft: ['pending'], pending: ['approved', 'aborted'], approved: ['executing', 'aborted'], executing: ['completed', 'aborted'], completed: [], aborted: [] },
	executionConfirmation: { pending: ['confirmed', 'failed'], confirmed: [], failed: ['cancelled'], cancelled: [] },
};

export function allowedTransitions(entity: string, status: string): readonly string[] {
	return ENTITY_TRANSITIONS[entity]?.[status] || [];
}

export const ENTITY_CONFIGS: readonly EntityConfig[] = [
  { key: 'reservoir', path: 'reservoirs', label: '库区', statuses: ['normal', 'warning', 'critical', 'restricted'] as const },
  { key: 'gateUnit', path: 'gates', label: '闸门', statuses: ['open', 'closed', 'moving', 'locked'] as const },
  { key: 'operationDirective', path: 'directives', label: '操作指令', statuses: ['draft', 'pending', 'approved', 'executing', 'completed', 'aborted'] as const },
  { key: 'executionConfirmation', path: 'confirmations', label: '执行确认', statuses: ['pending', 'confirmed', 'failed', 'cancelled'] as const }
];

File: backend/internal/constants/status.go
package constants

// Shared status values are mirrored in frontend/src/types/status.ts. Keeping
// the lists explicit makes state-machine drift visible during code review.

type GateState string

const (
	GateStateOpen   GateState = "open"
	GateStateClosed GateState = "closed"
	GateStateMoving GateState = "moving"
	GateStateLocked GateState = "locked"
)

var AllGateState = []string{"open", "closed", "moving", "locked"}

type DirectiveState string

const (
	DirectiveStateDraft     DirectiveState = "draft"
	DirectiveStatePending   DirectiveState = "pending"
	DirectiveStateApproved  DirectiveState = "approved"
	DirectiveStateExecuting DirectiveState = "executing"
	DirectiveStateCompleted DirectiveState = "completed"
	DirectiveStateAborted   DirectiveState = "aborted"
)

var AllDirectiveState = []string{"draft", "pending", "approved", "executing", "completed", "aborted"}

var ReservoirTransitions = map[string]map[string]bool{
	"normal":     {"warning": true, "critical": true},
	"warning":    {"critical": true, "restricted": true, "normal": true},
	"critical":   {"restricted": true, "warning": true},
	"restricted": {"critical": true},
}

var GateUnitTransitions = map[string]map[string]bool{
	"open":   {"moving": true, "locked": true},
	"closed": {"moving": true, "locked": true},
	"moving": {"open": true, "closed": true, "locked": true},
	"locked": {"closed": true},
}

var OperationDirectiveTransitions = map[string]map[string]bool{
	"draft":     {"pending": true},
	"pending":   {"approved": true, "aborted": true},
	"approved":  {"executing": true, "aborted": true},
	"executing": {"completed": true, "aborted": true},
	"completed": {},
	"aborted":   {},
}

var ExecutionConfirmationTransitions = map[string]map[string]bool{
	"pending":   {"confirmed": true, "failed": true},
	"confirmed": {},
	"failed":    {"cancelled": true},
	"cancelled": {},
}

func CanTransition(graph map[string]map[string]bool, from, to string) bool {
	targets, exists := graph[from]
	return exists && targets[to]
}
